x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/action/TransportPutLifecycleAction.java

```diff
@@ -100,13 +122,195 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                             logger.info("updating index lifecycle policy [{}]", request.getPolicy().getName());
                         }
                         IndexLifecycleMetadata newMetadata = new IndexLifecycleMetadata(newPolicies, currentMetadata.getOperationMode());
-                        newState.metaData(MetaData.builder(currentState.getMetaData())
+                        stateBuilder.metaData(MetaData.builder(currentState.getMetaData())
                                 .putCustom(IndexLifecycleMetadata.TYPE, newMetadata).build());
-                        return newState.build();
+                        ClusterState nonRefreshedState = stateBuilder.build();
+                        if (oldPolicy == null) {
+                            return nonRefreshedState;
+                        } else {
+                            try {
+                                return updateIndicesForPolicy(nonRefreshedState, xContentRegistry,
+                                    oldPolicy.getPolicy(), lifecyclePolicyMetadata);
+                            } catch (Exception e) {
+                                logger.warn(new ParameterizedMessage("unable to refresh indices phase JSON for updated policy [{}]",
+                                    oldPolicy.getName()), e);
+                                // Revert to the non-refreshed state
+                                return nonRefreshedState;
+                            }
+                        }
                     }
                 });
     }
 
+    /**
+     * Ensure that we have the minimum amount of metadata necessary to check for cache phase
+     * refresh. This includes:
+     * - An execution state
+     * - Existing phase definition JSON
+     * - A current step key
+     * - A current phase in the step key
+     * - Not currently in the ERROR step
```

Comment:
Why don't we allow the update of a policy if it's on the `ERROR` step? would it make sense to allow updating a policy even if on the error step? (I'm guessing this could be a good way to correct possible typos or such in the policy definition that are only caught at runtime - ie. `wait-for-snapshot` for a wrongly typed slm policy name)

Reply:
The ERROR step automatically refreshes the JSON when it's retried (either manually or automatically). I wanted not to refresh it on the ERROR step because otherwise the actual exception stored in the lifecycle execution state would reference a phase policy definition that could be a newer version than was executed (a user should be able to refer to the definition where the error occurred)

Reply:
Makes sense, thanks for the clarification